src/adagrams.js

```diff
@@ -1,15 +1,140 @@
+  R: 6,
+  S: 4,
+  T: 6,
+  U: 4,
+  V: 2,
+  W: 2,
+  X: 1,
+  Y: 2,
+  Z: 1,
+};
+
+
+function getRandomProperty() {
+  const keys = Object.keys(LETTER_POOL);
+
+  return keys[Math.floor(Math.random() * keys.length)];
+}
+
+let drawnLetters = [];
+
 export const drawLetters = () => {
-  // Implement this method for wave 1
+  while (drawnLetters.length < 10) {
+    let myCurrentKey = getRandomProperty();
+    let myCurrentValue = LETTER_POOL[myCurrentKey];
+    if (myCurrentValue !== 0) {
+      drawnLetters.push(myCurrentKey);
+      LETTER_POOL[myCurrentKey] = myCurrentValue - 1;
+    }
+    if (LETTER_POOL[myCurrentKey] === 0) {
+      delete LETTER_POOL[myCurrentKey];
+    }
+  }
+  
+  return drawnLetters;
 };
 
+
+
 export const usesAvailableLetters = (input, lettersInHand) => {
-  // Implement this method for wave 2
-};
+  let bankCopy = JSON.parse(JSON.stringify(lettersInHand))
+  for (let letter of input) {
+    if (!bankCopy.includes(letter.toUpperCase())){
+      return false
+    } else if (bankCopy.includes(letter.toUpperCase())) {
+      bankCopy = bankCopy.filter(value => value !== letter)
+      }
+    }
+    return true
+  };
+  
 
 export const scoreWord = (word) => {
-  // Implement this method for wave 3
-};
+  let score = 0;
+  let lengthOfWord = word.length;
+  const letterValueDict = { 
+     A: 1, E: 1, I: 1, O: 1, U: 1, L: 1,
+     N: 1, R: 1, S: 1, T: 1, D: 2, G: 2,
+     B: 3, C: 3, M: 3, P: 3, F: 4, V: 4,
+     W: 4, Y: 4, K: 5, J: 8, X: 8, Q: 10,
+     Z: 10, H:4
+    };
+
+    if (word.length === 0){
+      return 0;
+    }
+
+    if(lengthOfWord >= 7){
+      score += 8;
+    }
+
+    for (let i = 0; i < lengthOfWord; i++) {
+      const alpha = word[i].toUpperCase();
+      score += letterValueDict[alpha];
+    } 
+
+    return score;
+    
+}
+
 
 export const highestScoreFrom = (words) => {
-  // Implement this method for wave 4
+  let wordDict = {};
+  let listOfMaxKeys = [];
+  let maxValue;
+  
+  for (const word of words) {
+    wordDict[word] = scoreWord(word);
+  }
+
+  maxValue = Math.max(...Object.values(wordDict));
+  
+  
+  for (const [key, value] of Object.entries(wordDict)) {
+    if(value === maxValue) {
+      listOfMaxKeys.push(key);
+    }
+  }
+
+  if (listOfMaxKeys.length > 1) {
+    let tenLetterWordsList = [];
+    let nonTenLetterWordsList = [];
+
+    for (const word of listOfMaxKeys) {
+      if (word.length === 10) {
+        tenLetterWordsList.push(word);
+      } else {
+        nonTenLetterWordsList.push(word);
+      }
+    }
+    
+    if (tenLetterWordsList.length > 0) {
+      return { word: tenLetterWordsList[0], score: wordDict[tenLetterWordsList[0]] };
+    } else {
+      let sortedNonTenLetterWordList = nonTenLetterWordsList.sort((a, b) => a.length - b.length);
+      let shortestWord = sortedNonTenLetterWordList[0];
+      return { word: shortestWord, score: wordDict[shortestWord] };
+    }
+  }
+
+  return { word: listOfMaxKeys[0], score: wordDict[listOfMaxKeys[0]] };
 };
```

Comment:
I understand the concept behind what you are trying to do here! Using objects to separate out each step makes sense, but it starts to get tedious and more complex when we do that. Since we are returning an object with word and score as keys, why not create the one object that has the first word and the first word score as values? From there, we can run through the rest of the wordList and make the comparisons we need in place! This will drastically decrease the complexity and code you need to use! An example is below:

     export const highestScoreFrom = (words) => {
       let winningWord = {
         word: null,
         score: null
       }

       for (let word of words){

         let thisWord = {
           word: word,
           score: scoreWord(word)
         }

         if (thisWord.score > winningWord.score){
           winningWord = thisWord
         } else if (thisWord.score === winningWord.score){
           const isBestTen = winningWord.word.length === 10;
           const isWordTen = thisWord.word.length === 10;
           const isWordShorter = thisWord.word.length < winningWord.word.length;

           if ((isWordTen && !isBestTen) || (isWordShorter && !isBestTen)){
             winningWord = thisWord;
           }
         }
       }
       return winningWord;
     }